---
output:
  BiocStyle::html_document
---

<!---
The following chunk of code, which should not be shown in the resulting document (echo=FALSE)
sets up global processing options, such as forcing 'knitr' to stop when an error
in the R code is encountered, caching of the results in the 'cache'
directory and asking 'knitr' to figure out automatically the dependencies among
code chunks to re-calculate cached results (autodep=TRUE).

Other options could be changing the name of the directory where figures end up
('figure' by default), etc. For a full account of 'knitr' options please consult
http://yihui.name/knitr/options

At the end of the chunk a 'cat()' call is made to dump a CSS file that gives
a better look-and-feel than the knitr default one. See the source css/ieo.css
and the resulting projectTemplate.html to understand where this is being dumpted.
--->


```{r setup, cache=FALSE, echo=FALSE, results='hide', message=FALSE}
library(knitr)

opts_chunk$set(cache=TRUE,
               cache.path="cache/QA",
               cache.extra=R.version.string,
               autodep=TRUE,
               fig.align="center",
               comment="")
```

# Quality assessment

## Data import

We start importing the raw table of counts.

<!--
The option 'message=FALSE' avoid dumping R messages such as "Loading required package: methods"
into the output of the report.
-->

```{r, message=FALSE}
library(SummarizedExperiment)
setwd("/home/eva/Bioinf/term3/IEO/project(I)/projectTemplate")
se <- readRDS(file.path("rawCounts", "seHNSC.rds"))
se
```

Explore the column (phenotypic) data, which in this case corresponds to clinical
variables, and their corresponding metadata.

```{r}
dim(colData(se))
colData(se)[1:5, 1:5]
mcols(colData(se), use.names=TRUE)
```

These metadata consists of two columns of information about the clinical variables.
One called `labelDescription` contains a succint description of the variable, often
not more self-explanatory than the variable name itself, and the other called
'CDEID' corresponds to the so-called `Common Data Element (CDE)` identifier. This
identifier can be use in https://cdebrowser.nci.nih.gov to search for further
information about the associated clinical variable using the `Advanced search`
form and the `Public ID` attribute search.

Now, explore the row (feature) data.

```{r}
rowData(se)
rowRanges(se)
```

To perform quality assessment and normalization we need first to load the
[edgeR](http://bioconductor.org/packages/edgeR) R/Bioconductor package and
create a `DGEList' object.

```{r, message=FALSE}
library(edgeR)

dge <- DGEList(counts=assays(se)$counts, genes=mcols(se))
saveRDS(dge, file.path("results", "dge.rds"))
```

Now calculate $\log_2$ CPM values of expression and put them as an additional
assay element to ease their manipulation.

```{r}
assays(se)$logCPM <- cpm(dge, log=TRUE, prior.count=0.5)
assays(se)$logCPM[1:5, 1:5]
```

## Sequencing depth

Let's examine the sequencing depth in terms of total number of sequence read counts
mapped to the genome per sample. Figure \@ref(fig:libsizes) below shows the
sequencing depth per sample, also known as library sizes, in increasing order.

<!---
you can control the height and width in pixels of the figure with 'out.height' and
'out.width'. Figures are automatically numbered, to refer to them in the main test
you should use the notation shown above as \@ref(fig:xxxx) with xxxx being the label
in the code chunk that also gives the filename of the figure. This name must be unique
--->

```{r libsizes, echo=FALSE, out.width="600px", fig.cap="Library sizes in increasing order."}
ord <- order(dge$sample$lib.size/1e6)
barplot(dge$sample$lib.size[ord]/1e6, las=1, ylab="Millions of reads", border=NA,
                xlab="Samples", col=c("blue", "red")[(se$type[ord] == "tumor") + 1])
legend("topleft", c("tumor", "normal"), fill=c("red", "blue"), inset=0.01)
```
This figure reveals substantial differences in sequencing depth between samples
and we may consider discarding those samples whose depth is substantially lower
than the rest. To identify who are these samples we may simply look at the
actual numbers including portion of the sample identifier that distinguishes them.

```{r}
sampledepth <- round(dge$sample$lib.size / 1e6, digits=1)
names(sampledepth) <- substr(colnames(se), 6, 12)
sort(sampledepth)
```


```{r}
dge <- dge[,(dge$samples$lib.size/1e6) > 40 ] # 50e06 threshold value
se <- se[,rownames(dge$samples)]
table(se$type)
```

## Distribution of expression levels among samples

Let's look at the distribution of expression values per sample in terms of
logarithmic CPM units. Due to the large number of samples, we display tumor
and normal samples separately, and are shown in Figure \@ref(fig:distRawExp)

```{r distRawExp, echo=FALSE, fig.height=4, fig.width=10, out.width="800px", fig.cap="Non-parametric density distribution of expression profiles per sample.", message=FALSE, eval=FALSE}
library(geneplotter)
par(mfrow=c(1, 2))
multidensity(as.list(as.data.frame(assays(se[, se$type == "tumor"])$logCPM)),
                          xlab="log 2 CPM", legend=NULL, main="Tumor samples", las=1)
multidensity(as.list(as.data.frame(assays(se[, se$type == "normal"])$logCPM)),
                          xlab="log 2 CPM", legend=NULL, main="Normal samples", las=1)
```

We do not appreciate substantial differences between the samples in the
distribution of expression values.

## Distribution of expression levels among genes

Let's calculate now the average expression per gene through all the samples.
Figure \@ref(fig:exprdist) shows the distribution of those values across genes.

```{r exprdist, echo=FALSE, out.width="400px", fig.cap="Distribution of average expression level per gene."}
avgexp <- rowMeans(assays(se)$logCPM)
hist(avgexp, xlab="log2 CPM", main="", las=1)
abline(v=1, col="red", lwd=2)
```

## Filtering of lowly-expressed genes

In the light of this plot, we may consider a cutoff of 1 log CPM unit as minimum value
of expression to select genes being expressed across samples. Using this cutoff we proceed
to filter out lowly-expressed genes.

```{r}
mask <- avgexp > 1
dim(se)
se.filt <- se[mask, ]
dim(se.filt)
dge.filt <- dge[mask, ]
dim(dge.filt)
```

Store un-normalized versions of the filtered expression data.

```{r}
saveRDS(se.filt, file.path("results", "se.filt.unnorm.rds"))
saveRDS(dge.filt, file.path("results", "dge.filt.unnorm.rds"))
```

## Normalization

We calculate now the normalization factors on the filtered expression data set.

```{r}
dge.filt <- calcNormFactors(dge.filt)
```

Replace the raw log2 CPM units in the corresponding assay element of the `SummarizedExperiment`
object, by the normalized ones.

```{r}
assays(se.filt)$logCPM <- cpm(dge.filt, log=TRUE, normalized.lib.sizes=TRUE, prior.count=0.25)
```

Store normalized versions of the filtered expression data.

```{r}
saveRDS(se.filt, file.path("results", "se.filt.rds"))
saveRDS(dge.filt, file.path("results", "dge.filt.rds"))
```

## MA-plots

We examine now the MA-plots of the normalized expression profiles. We look first to
the tumor samples in Figure \@ref(fig:maPlotsTumor).

```{r maPlotsTumor, fig.height=36, fig.width=6, dpi=100, echo=FALSE, fig.cap="MA-plots of the tumor samples."}
par(mfrow=c(22, 3), mar=c(4, 5, 3, 1))
setmp <- se.filt[, se.filt$type == "tumor"]
dgetmp <- dge.filt[, se.filt$type == "tumor"]
for (i in 1:ncol(setmp)) {
  A <- rowMeans(assays(setmp)$logCPM)
  M <- assays(setmp)$logCPM[, i] - A
  samplename <- substr(as.character(setmp$bcr_patient_barcode[i]), 1, 12)
  smoothScatter(A, M, main=samplename, las=1)
  abline(h=0, col="blue", lwd=2)
  lo <- lowess(M ~ A)
  lines(lo$x, lo$y, col="red", lwd=2)
}
```

We do not observe samples with major expression-level dependent biases. Let's
look now to the normal samples in Figure \@ref(fig:maPlotsNormal).

```{r maPlotsNormal, fig.height=18, fig.width=6, dpi=100, echo=FALSE, fig.cap="MA-plots of the normal samples."}
par(mfrow=c(9, 3), mar=c(4, 5, 3, 1))
setmp <- se.filt[, se.filt$type == "normal"]
dgetmp <- dge.filt[, se$type == "normal"]
for (i in 1:ncol(setmp)) {
  A <- rowMeans(assays(setmp)$logCPM)
  M <- assays(setmp)$logCPM[, i] - A
  samplename <- substr(as.character(setmp$bcr_patient_barcode[i]), 1, 12)
  smoothScatter(A, M, main=samplename, las=1)
  abline(h=0, col="blue", lwd=2)
  lo <- lowess(M ~ A)
  lines(lo$x, lo$y, col="red", lwd=2)
}
```

We do not observe either important expression-level dependent biases among the normal samples.

## Batch identification

We will search now for potential surrogate of batch effect indicators. Given that each sample
names corresponds to a TCGA barcode (see https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode),
following the strategy described in http://bioinformatics.mdanderson.org/main/TCGABatchEffects:Overview
we are going to derive different elements of the TCGA barcode and examine their distribution
across samples.

```{r}
tss <- substr(colnames(se.filt), 6, 7)
table(tss)
center <- substr(colnames(se.filt), 27, 28)
table(center)
plate <- substr(colnames(se.filt), 22, 25)
table(plate)
portionanalyte <- substr(colnames(se.filt), 18, 20)
table(portionanalyte)
samplevial <- substr(colnames(se.filt), 14, 16)
table(samplevial)
```

From this information we can make the following observations:

  * 

  *

  *

  * 

  * 
We are going to use the TSS as surrogate of batch effect indicator. Considering our outcome
of interest as molecular changes between sample types, tumor vs. normal, we will examine now
the cross-classification of this outcome with TSS.

```{r}
table(data.frame(TYPE=se.filt$type, TSS=tss))
plate = substr(colnames(se.filt), 22, 25)
table(data.frame(TYPE=se.filt$type, PLATE=plate))
portionanalyte <- substr(colnames(se.filt), 18, 20)
table(data.frame(TYPE=se.filt$type, PORTION=portionanalyte))
samplevial <- substr(colnames(se.filt), 14, 16)
table(data.frame(TYPE=se.filt$type, VIAL=samplevial))

```

#### TSS

```{r sampleClustering, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Hierarchical clustering of the samples."}

logCPM <- cpm(dge.filt, log=TRUE, prior.count=3)
d <- as.dist(1-cor(logCPM, method="spearman"))
sampleClustering <- hclust(d)
batch <- as.integer(factor(tss))
sampleDendrogram <- as.dendrogram(sampleClustering, hang=-1, label_cols = manualcolors)
names(batch) <- colnames(se.filt)
outcome2 <- as.character(se.filt$type)
outcome2 <- replace(outcome2, outcome2 == 'tumor', "T")
outcome2 <- replace(outcome2, outcome2 == 'normal', "N")
names(outcome2) <- colnames(se.filt)
colors <- palette()
palette(c(colors,'black','forestgreen', 'red2', 'orange', 'cornflowerblue', 
                'magenta', 'darkolivegreen4',  
                'indianred1', 'tan4', 'darkblue', 
                'mediumorchid1','firebrick4',  'yellowgreen', 'lightsalmon', 'tan3',
                "tan1",'darkgray', 'wheat4', '#DDAD4B', 'chartreuse', 'seagreen1',
                'moccasin', 'mediumvioletred', 'seagreen','cadetblue1',
                "darkolivegreen1" ,"tan2" ,   "tomato3"))
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome2)

plot(sampleDendrogram, cex = 0.6 , main="Hierarchical clustering of samples")
legend("topright", paste("Batch", sort(unique(batch)), levels(factor(tss))), fill=unique(as.factor(tss)))
```
It is difficult to see much in this plot, but in general, we notice that batches seem mixed.

```{r mdsPlot, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples."}
outcome2 <- replace(outcome2, outcome2 == 'tumor', "T")
outcome2 <- replace(outcome2, outcome2 == 'normal', "N")
batch <- as.integer(factor(tss))
plotMDS(dge.filt, labels=outcome2, col=batch)
legend("bottomleft", paste("Batch", sort(unique(batch)), levels(factor(tss))),
       fill=sort(unique(batch)), inset=0.05)
```

### Removal of samples with uneven distribution.

We remove all the samples from tss that lack normal or tumor samples. We only consider TSS "CV", "H7", "HD", "WA".
```{r}
mask_tss <- tss %in% c("CV", "H7", "HD", "WA")
se_tss <- se.filt[,mask_tss]
tss_filt <- substr(colnames(se_tss), 6, 7)
table(data.frame(TYPE=se_tss$type, TSS=tss_filt))
dge_tss <- dge.filt[,mask_tss]
dim(dge_tss)
```
```{r sampleClustering2, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Hierarchical clustering of the samples after removing those from unbalanced tissue sites."}

logCPM <- cpm(dge_tss, log=TRUE, prior.count=3)
d <- as.dist(1-cor(logCPM, method="spearman"))
sampleClustering <- hclust(d)
batch <- as.integer(factor(tss_filt))
sampleDendrogram <- as.dendrogram(sampleClustering, hang=-1)
names(batch) <- colnames(se_tss)
outcome3 <- as.character(se_tss$type)
outcome3 <- replace(outcome3, outcome3 == 'tumor', "T")
outcome3 <- replace(outcome3, outcome3 == 'normal', "N")
names(outcome3) <- colnames(se_tss)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome3)

plot(sampleDendrogram, cex = 0.6 , main="")
legend("topright", paste("Batch", sort(unique(batch)), levels(factor(tss_filt))), fill=unique(as.factor(batch)))

```
Batches are mixed and we observed N and T clustering. There are two clusters of N that could be due to othervariables.

```{r mdsPlot2, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples after removing those from unbalanced tissue sites."}
plotMDS(dge_tss, label = outcome3, col=batch, main="Clustering by tss")
legend("bottomleft", paste("Batch", sort(unique(batch)), levels(factor(tss_filt))),
       fill=sort(unique(batch)), inset=0.05)
```

### Checking other technical posible surrogate variables.

```{r}
plate_filt <- substr(colnames(se_tss), 22, 25)
table(plate_filt)
portionanalyte_filt <- substr(colnames(se_tss), 18, 20)
table(portionanalyte_filt)
samplevial_filt <- substr(colnames(se_tss), 14, 16)
table(samplevial_filt)
```

Since PCA seem to be more readable, we only plot PCA:

```{r mdsPlot3, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples colored by plate."}
batch <- as.integer(factor(plate_filt))
plotMDS(dge_tss, label = outcome3, col=batch, main = "")
legend("bottomleft", paste("Batch", sort(unique(batch)), levels(factor(plate_filt))),
       fill=sort(unique(batch)), inset=0.05)
```

```{r mdsPlot4, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples colored by portionanalyte."}
batch <- as.integer(factor(portionanalyte_filt))
plotMDS(dge_tss, label = outcome3, col=batch, main = "")
legend("bottomleft", paste("Batch", sort(unique(batch)), levels(factor(portionanalyte_filt))),
       fill=sort(unique(batch)), inset=0.05)
```


```{r mdsPlot5, fig.height=7, fig.width=14, dpi=100, echo=TRUE, fig.cap="Multidimensional scaling plot of the samples colored by samplevial."}

batch <- as.integer(factor(samplevial_filt))
plotMDS(dge_tss, label = outcome3, col=batch, main = "")
legend("bottomleft", paste("Batch", sort(unique(batch)), levels(factor(samplevial_filt))),
       fill=sort(unique(batch)), inset=0.05)
```

## Looking for surrogate variables

```{r}
library(sva)
mod <- model.matrix(~type , data = colData(se_tss))
head(mod)
```

Null model.

```{r}
mod0 <- model.matrix(~ 1, data = colData(se_tss))
```
We estimate the surrogate variables calling the sva() function as follows:

```{r}
sv <- sva(logCPM, mod, mod0)
```


Quick overview of the impact of adjusting for the estimated heterogeneity using `sva` F-test.
First, we examine first how many genes change across conditions without adjustment. It is not
the most the most correct way to do this, since we are using normalized log 2 CPM values, which have an
underlying relationship between mean and variance, but it will still be useful for comparing the model with and without surrogate variables.

```{r}
pValues <- f.pvalue(logCPM, mod, mod0)
sum(p.adjust(pValues, method = "BH") < 0.05)
```

```{r}
hist(pValues, main = "Before adjustment for surrogate variables", las = 1)
```

Now we examine the number of changes when we adjust for the estimated surrogate variables:

```{r}
modSv <- cbind(mod, sv$sv)
mod0Sv <- cbind(mod0, sv$sv)
pValuesSv <- f.pvalue(logCPM, modSv, mod0Sv)
sum(p.adjust(pValuesSv, method = "BH") < 0.05)
```

The number of differently expressed genes has increased by about a 19%. 

```{r}
#par(mfrow=c(1, 2))
hist(pValues, col=rgb(0,0,1,1/4),main = "Before adjustment for surrogate variables", las = 1, ylim = c(0,9000), breaks = 40)
hist(pValuesSv,  col=rgb(1,0,0,1/4), main = "After adjustment for surrogate variables", las = 1, ylim = c(0,9000), add=T, breaks = 40)
legend( 0.6,6000, c("Before adjustment", "After adjustment"), fill=c(rgb(0,0,1,1/4), rgb(1,0,0,1/4)))
```

```{r}
par(mfrow=c(1, 2))
hist(pValues,main = "Before adjustment for surrogate variables", las = 1, ylim = c(0,9000), breaks = 40)
hist(pValuesSv,  main = "After adjustment for surrogate variables", las = 1, ylim = c(0,9000), breaks = 40)
```


The histogram of p-values has improved after adjustment.


## Session information

```{r, message=FALSE}
sessionInfo()
```
